### PR TITLE
Update rust vesion to 1.57

### PR DESCRIPTION
**Pull request by xudong963**
# Which issue does this PR close?

Related: https://blog.rust-lang.org/2021/12/02/Rust-1.57.0.html

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by houqp** [comment]
@xudong963 do you want to also fix the clippy errors?

**Comment by xudong963** [comment]
> @xudong963 do you want to also fix the clippy errors?

Sure 😄, a little late.

File: datafusion/src/scalar.rs [hunk 1]
```diff
@@ -68,7 +68,7 @@ pub enum ScalarValue {
     /// large binary
     LargeBinary(Option<Vec<u8>>),
     /// list of nested ScalarValue (boxed to reduce size_of(ScalarValue))
-    #[allow(clippy::box_vec)]
+    #[allow(clippy::box_collection)]
```
Comment (xudong963) [inline datafusion/src/scalar.rs:71]: https://rust-lang.github.io/rust-clippy/master/index.html#box_collection

File: datafusion/src/physical_plan/datetime_expressions.rs [hunk 2]
```diff
@@ -77,7 +78,10 @@ where
         })?;
 
     // first map is the iterator, second is for the `Option<_>`
-    array.iter().map(|x| x.map(|x| op(x)).transpose()).collect()
+    array
```
Comment (xudong963) [inline datafusion/src/physical_plan/datetime_expressions.rs:81]: https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure

File: datafusion/src/sql/parser.rs [hunk 3]
```diff
@@ -85,7 +85,7 @@ pub struct CreateExternalTable {
 #[derive(Debug, Clone, PartialEq)]
 pub enum Statement {
     /// ANSI SQL AST node
-    Statement(SQLStatement),
+    Statement(Box<SQLStatement>),
```
Comment (xudong963) [inline datafusion/src/sql/parser.rs:88]: https://rust-lang.github.io/rust-clippy/master/index.html#large_enum_variant

**Comment by xudong963** [comment]
I can't find `field is never read` clippy update in https://github.com/rust-lang/rust-clippy/blob/master/CHANGELOG.md#rust-157

8KxCTexjiN

So I still don't fix clippy errors in the screenshot.

**Review by alamb — COMMENTED** [review]
Thank you @xudong963  -- there are still a few clippy failures https://github.com/apache/arrow-datafusion/runs/4406965470?check_suite_focus=true

They mostly look related to "fields not read" and some of that is surprising to me (like perhaps we have some missing functionality that needs to be fleshed out). 

One approach we might take is add `clippy allow` lints and then go work through them one by one. I'll try and do this later today if I get to it

**Comment by xudong963** [comment]
> One approach we might take is add `clippy allow` lints and then go work through them one by one. I'll try and do this later today if I get to it

You mean add `#[allow(dead_code)]`?

**Comment by alamb** [comment]
@xudong963  I pushed a few changes to this branch to try and get clippy to pass. Here's hoping we get a clean run

File: datafusion/src/physical_plan/expressions/average.rs [hunk 4]
```diff
@@ -37,8 +37,8 @@ use super::{format_state_name, sum};
 #[derive(Debug)]
 pub struct Avg {
     name: String,
+    #[allow(dead_code)]
```
Comment (alamb) [inline datafusion/src/physical_plan/expressions/average.rs:40]: I will look into removing this field as well

File: datafusion/src/physical_plan/windows/built_in.rs [hunk 5]
```diff
@@ -33,28 +31,22 @@ use std::sync::Arc;
...
     expr: Arc<dyn BuiltInWindowFunctionExpr>,
     partition_by: Vec<Arc<dyn PhysicalExpr>>,
     order_by: Vec<PhysicalSortExpr>,
-    window_frame: Option<WindowFrame>,
 }
 
 impl BuiltInWindowExpr {
     /// create a new built-in window function expression
     pub(super) fn new(
-        fun: BuiltInWindowFunction,
```
Comment (alamb) [inline datafusion/src/physical_plan/windows/built_in.rs:46]: the fun and window_frame are used to construct the `BuiltInWindowFunctionExpr` (so the don't need to be also encoded on the `BuildInWindowExpr`

File: benchmarks/src/bin/tpch.rs [hunk 6]
```diff
@@ -75,10 +75,9 @@ struct BallistaBenchmarkOpt {
     #[structopt(short = "i", long = "iterations", default_value = "3")]
     iterations: usize,
 
-    /// Batch size when reading CSV or Parquet files
-    #[structopt(short = "s", long = "batch-size", default_value = "8192")]
-    batch_size: usize,
-
+    // /// Batch size when reading CSV or Parquet files
```
Comment (alamb) [inline benchmarks/src/bin/tpch.rs:78]: These parameters are never read (aka they don't do anything) so removing them from the CLI seemed like a reasonable thing to do

File: ballista/rust/core/src/config.rs [hunk 7]
```diff
@@ -31,22 +31,22 @@ pub const BALLISTA_DEFAULT_SHUFFLE_PARTITIONS: &str = "ballista.shuffle.partitio
 #[derive(Debug, Clone)]
 pub struct ConfigEntry {
     name: String,
-    description: String,
-    data_type: DataType,
+    _description: String,
```
[the remark this answers is not in the document]
Comment (xudong963) [reply ballista/rust/core/src/config.rs:34]: Yes, very much agree with you, `_` is a conservative but useful way!

**Comment by xudong963** [comment]
Nice, clippy passed😄

File: datafusion/src/physical_plan/expressions/average.rs (hunk 4 above)
Comment (capkurmagati) [reply datafusion/src/physical_plan/expressions/average.rs:40]: This PR might be the reason causes the different behavior on `dead_code`. 
https://github.com/rust-lang/rust/pull/85200

**Review by alamb — APPROVED** [review]

File: datafusion/src/physical_plan/expressions/average.rs [hunk 8]
```diff
@@ -69,10 +67,14 @@ impl Avg {
         name: impl Into<String>,
         data_type: DataType,
     ) -> Self {
+        // Average is always Float64, but Avg::new() has a data_type
+        // parameter to keep a consistent signature with the other
+        // Aggregate expressions.
+        assert_eq!(data_type, DataType::Float64);
```
Comment (xudong963) [inline datafusion/src/physical_plan/expressions/average.rs:73]: 👍

File: datafusion/src/physical_plan/expressions/average.rs (hunk 4 above)
Comment (alamb) [reply datafusion/src/physical_plan/expressions/average.rs]: I think you are right

**Review by houqp — APPROVED** [review]

**Comment by alamb** [comment]
Thanks again @xudong963.

**Pull request merged**